### DISCUSSION TITLE
[CW-2008] feat: Adding VOIP Integration/Support

### DISCUSSION THREAD

**Discussion by jberenback** [chatwoot/chatwoot#8451] [Feature Requests]
One of the reasons our team is hesitant to consider Chatwoot as we look for something other than Zendesk is the fact we can not support VOIP. Since we are a website that does rely on phone calls (much less than a call center), we need VOIP. If we were to migrate to a new system, all of these items (texting, voice, email, chat etc), would need to be encompassed under one umbrella so that we wouldn't have disjointed systems. 

Right now we use Zendesk and Zipwhip. But Zendesk is clunky when it comes to most things and being able to have our product team integrate everything without our internal site would allow us to have a much better system overall. Just a thought as I'm sure many who are looking to move to something self hosted require VOIP (if they are using a help desk that has it), and would need it with the new system, rather than with another app. 

Thanks!

<sub>[CW-2008](https://linear.app/chatwoot/issue/CW-2008/feat-adding-voip-integrationsupport)</sub>

**Comment by iammattmartin** [comment]
This would be good long term but I would imagine it would be very far down the list given that chatwoot is still quite a new product and some core functions still need adding in...

Even long term SaaS offerings have only just added this and they've been in dev 10 + years.

**Comment by jcmatoskx** [comment]
Would love to see VOIP support as well, being able to have on customer history whenever they contacted us or were contacted by phone would be amazing, as well as being able to know who is the customer who is calling us and immediately be able to easily look at their contact history.

**Comment by Apesy** [comment]
VoIP, Hold Music, IVR, Customer History, all encompassed under Chatwoot would make definitely make this a good alternative to Zendesk. Unfortunately due to COVID-19 having Omnichannel Support Solutions such as Zendesk, Zoho, Freshdesk, etc, have become unaffordable due to their ridiculous high prices. If Chatwoot could implement most features that these other major companies have and STILL remain open-source this would be the perfect solution that most companies would switch too. I mean for Zendesk, you're looking at roughly $4,000 a month per agent for an enterprise sized solution.

**Comment by andredanner** [comment]
I have a **startup running that does chat- and voicebots** for specific sme use cases. The only reason that I am currently reluctant to offer my customers chatwoot as fully flechted contact center approach including bot integration is the missing VOIP Part.

More than happy to follow the discussion and/or to start into the design discussion. As I am not really feeling familiar with the chatwoot architectur and code base, I am a bit reluctant to offer implementation support. Perhaps, if someone could guide me ....

Best Regards

Andre

**Comment by jatosmaime** [comment]
VOIP is the only missing part!

**Comment by luissakihama** [comment]
Yes we would love that! But I was thinking, that maybe a combination of API Inbox with Dashboard App could  be a workaround solution for this. API Inbox could be integrated with Asterisk via REST to trigger calls and get status, and deploying an Dashboard App to receive / make calls via WebRTC. Is that make sense?

**Comment by egejo** [comment]
Agree VOIP is a missing important link...Love to see it on chatwoot

**Comment by tobsowo** [comment]
VoIP will be great feature to add to Chatwoot as it will make it stand out and more popular to enterprise grade solutions.

**Comment by clairton** [comment]
+1

**Comment by sdekna** [comment]
I think VOIP these days is becoming a must for any customer engagement app... +++1
Also, with the bursting numbers of open source voip solutions out there, it shouldn't be a big development commitment to allow it to hook up with some/any of these services.
A simple in-house solution would be something like webRTC... or an integration plugin to any of the other providers out there.

**Comment by wokawoka** [comment]
+1

**Comment by eremeye** [comment]
+1

**Comment by jmelati** [comment]
+1

**Comment by luiszxcv** [comment]
+1

**Comment by Aghiles97** [comment]
This feature is definitely needed and requested a lot, I hope in the near future we will see this in action.

**Comment by caribetodoincluido** [comment]
Facebook leads + **Chatwoot** (Voip + Dialer + ACD + WhatsApp Api (Forms) + Catalog + Payments ) 
Google ADS (Call Campaign) + **Chatwoot** (Voip + Dialer + ACD + WhatsApp Api (Forms) + Catalog + Payments ) 
Click to Call (Web Widget+Social Networks) + **Chatwoot** (Voip + Dialer + ACD + WhatsApp Api (Forms) + Catalog + Payments ) 

Each Traveler needs a phone call for urgent inquiry during a trip or just for take a relax vacations knowing that your travel concierge (travel business) is so close as a call. Is a priority channel for several business.

🤑🤑🤑 🚀

**Comment by Aghiles97** [comment]
> Facebook leads + **Chatwoot** (Voip + Dialer + ACD + WhatsApp Api (Forms) + Catalog + Payments ) Google ADS (Call Campaign) + **Chatwoot** (Voip + Dialer + ACD + WhatsApp Api (Forms) + Catalog + Payments ) Click to Call (Web Widget+Social Networks) + **Chatwoot** (Voip + Dialer + ACD + WhatsApp Api (Forms) + Catalog + Payments )
> 
> Each Traveler needs a phone call for urgent inquiry during a trip or just for take a relax vacations knowing that your travel concierge (travel business) is so close as a call. Is a priority channel for several business.
> 
> money_mouth_facemoney_mouth_facemoney_mouth_face rocket

Not only that, but Chatwoot can become a rich complete solution that can surpass popular CRM that already supports VOIP integration with popular PBX technologies like Asterisk.

**Comment by LiLiJal** [comment]
Hey  @sojan-official  is VOIP integration somewhere in Chatwoot  roadmap  based on these huge interest ?

**Comment by luiszxcv** [comment]
É realmente um recurso necessário.

**Comment by PP-Contrib** [comment]
I can help on the asterisk / FreePBX part

**Comment by dd0ke** [comment]
Indeed the feature would be beautiful. I would suggest to stay neutral vis-à-vis the PBX, and work on the SIP aspects only. Many WebPHones, javascript libraries and such offer WebRTC support, and that allows a vendor generic solution. 

Congratulations for such high values, Chatwoot seems to be really a stand-out in Open Source offerings!

**Comment by dd0ke** [comment]
There is a pretty straight forward integration I can suggest, to start with. It does not use a WebPhone, but rather provides a way to bridge caller and agent as soon as the agent gets assigned a conversation.
I'm happy to provide the server-side playground for the telephony components for the integration dev.

Proposed MVP:
1. Remain vendor agnostic, and use SIP capabilities:
  * The inbound call routing and logic can collect initial data from the caller, besides the Caller ID. An interactive Voice response can ask for language preference, account number, tracking number, reservation code, etc, and place the caller in the appropriate queue. This would be done per anyone's preference and tools, be it Asterisk, FreeSWITCH, Microsoft Teams, Twilio, etc.
  * The values collected are carried on to Chatwoot using X-Headers in the SIP invite. (Flags such as 'X-User-Authenticated: true' can let the agent know the user has already been identified via account number and PIN, for example)
2. Chatwoot could provide dynamic API calls for the PBX to query, with parameters such as current expected hold time, choice of music on hold, custom announcements, SIP URI for the CSAT (where caller is transferred after a conversation ends), unique CallBack URL for the conversation, etc.
3. Chatwood would permit user account parameters for VoIP settings, including the extension or device the agent uses to receive calls, any necessary authentication information, etc.
4. On a live view of conversations and inbound call flows, an agent can pick a call to get assigned, which then triggers the caller to be forwarded to the agent's extension, as defined in the user's settings.
5. Thereafter, the status of agent switches to busy, until the call ends and the PBX sends a status report to the specified CallBack URL, which updates the conversation.

Going forward, additional capabilities could be added, such as skills based routing, which could already be done somewhat via the X-Headers idea. Ideally, the inbound call flows should show the Caller ID, and Called Number, at a minimum. The Caller ID is to lookup any existing contacts for the caller, and the Called Number can be used to lookup what company, department, etc, is trying to be reached. 

Future implementations can include a SIP user-agent for Chatwoot that stays in the call flow at all times, as if every call was a conference call (caller, agent, Chatwoot Bot). This could permit offering benefits like agent supervision, Chatwoot side call recording and transcription, transferring calls from one agent to the other from the Chatwoot interface, and just live session handling options.

**Comment by dd0ke** [comment]
If a WebPhone was required, there are plenty. I will try this one shortly: https://github.com/gmaruzz/saraphone

**Comment by kj2222** [comment]
@ekavun @PPaiva can we collaborate on this?

**Comment by dd0ke** [comment]
I’m in! Let me know where we should move the conversation and get organized.

**Comment by PP-Contrib** [comment]
Also in. :)

**Comment by kj2222** [comment]
@ekavun @PPaiva Let's move this to email. Just drop me your email address here: https://forms.gle/YjrsxLvmRJovmNkz6 and let's get started

In case @LiLiJal @jcmatoskx @andredanner @Aghiles97 or someone else wants to collaborate on this

**Reply by Evoltstore** [reply]
The link gives "You need permission
This form can only be viewed by users in the owner's organisation.

Try contacting the owner of the form if you think that this is a mistake"

**Reply by kj2222** [reply]
Thanks for letting me know. Settings are changed now please check

**Comment by dd0ke** [comment]
Hi!

The link gives:
"This form can only be viewed by users in the owner's organization."

Can’t access it.

Cheers!

**Reply by Aghiles97** [reply]
same here

**Reply by kj2222** [reply]
Changed settings now please check

**Comment by colpari** [comment]
Hey folks, is there any move on this? We'd like to participate and would also to weigh in resources on this.

**Comment by colpari** [comment]
See also here: https://github.com/orgs/chatwoot/discussions/9437#discussion-6627830

**Comment by harismateen** [comment]
Eagerly waiting for this VOIP feature

**Comment by kj2222** [comment]
Hi everyone so after a really long time we've finally integrated voice. Right now it's a simple click to call option with freeswitch doing all the heavy lifting. Happy to guide others in this. Unfortunately we forked long ago and our push won't be compatible with current versions of chatwoot but we'd be glad to offer our inputs if someone wants to make an upstream push

**Reply by farverio** [reply]
@kj2222 do you have a repo you can share?

**Reply by badereddineqodia** [reply]
Hello @kj2222,
Would you mind if you share with us the link of the repo, please ?

**Comment by dd0ke** [comment]
Great! Do you have some kind of event based ticketing for inbound callers? Where do we find your code? I’ll see to make it current on recent FreesWITCH, but I’m a bit allergic to open source perversion to commercial. I do encourage their funding and opening up to bigger opportunities, and love all their team. There is a lot that can be done just, and only, observing SIP standards.

**Reply by kj2222** [reply]
It's currently mostly focused on outbound. So a click to dial where the agent's number is dialled first and then the customer's number and then bridged. PSTN is more important for us so we haven't focused on WebRTC or a softphone. For inbound calls, just based on the phone number create a private note of the call record in the user's most recent conversation. Will eventually also route the call to the agent who was assigned the call. Like I mentioned, all the heavy lifting is on the Freeswitch side.

Also apologies I meant current version of chatwoot not Freeswitch

**Reply by tobsowo** [reply]
I'd like to learn about Freeswitch and PSTN, if there's any resources you can share?

**Comment by tech-mint** [comment]
Dear Chatwoot Development Team,

I hope this message finds you well. I am writing to propose a potential integration between your outstanding customer support platform, Chatwoot, and the Issabel Asterisk telephony system. As a robust open-source PBX solution, Issabel Asterisk is widely adopted in various industries, providing efficient and scalable communication solutions.

Integrating Issabel Asterisk with Chatwoot would bring significant value to both platforms and their users. Here’s why:

Streamlined Communication: By connecting voice interactions from Issabel Asterisk directly into Chatwoot, customer support teams can manage all communication channels—voice, chat, email, and social media—from a single unified interface. This would reduce response times and improve the efficiency of support agents.

Improved Customer Experience: Having all customer interactions in one place allows for better context and continuity in support conversations. Agents can quickly access previous interactions and provide a more personalized and effective customer service experience.

Enhanced Analytics and Reporting: An integrated system would allow businesses to gather and analyze data across all communication channels. This comprehensive view would empower managers to make data-driven decisions to improve team performance and customer satisfaction.

Cost Efficiency: The integration would help companies that already use Issabel Asterisk to leverage their existing infrastructure while maximizing the capabilities of Chatwoot, reducing the need for additional tools and training.

Competitive Advantage: Offering this integration would attract a broader range of customers who are looking for a versatile and comprehensive customer communication platform. It would also strengthen Chatwoot’s position in the market as a solution that adapts to the diverse needs of modern businesses.

I believe that the synergy between Chatwoot and Issabel Asterisk could bring considerable advantages to businesses seeking to elevate their customer support operations. I would be happy to provide more details or assist in any way possible to facilitate this integration.

Thank you for considering this proposal. I look forward to hearing your thoughts on this exciting opportunity.

**Comment by Juandaaguirre** [comment]
+1

**Comment by ssshojaei** [comment]
+1

**Comment by mansourmahboubi** [comment]
+1

**Comment by gomes0202** [comment]
+1

**Comment by abehnamfard** [comment]
+1

**Comment by counthavoc** [comment]
+1

**Comment by tuhinmallick** [comment]
+1

**Comment by MeatBoyed** [comment]
+1 ****

**Comment by badereddineqodia** [comment]
+++++++++++++1

**Comment by ldaniel-jmz** [comment]
+1

**Comment by dealmengor** [comment]
+++++++++ 20!

**Comment by sojan-official** [comment]
The team is currently working on a v1 for this : https://github.com/chatwoot/chatwoot/issues/11481

**Reply by jcmatoskx** [reply]
Awesome to know! thanks for listening to the community. Loving the development pace of Chatwoot lately!

**Reply by AnxHall** [reply]
LET'S GO !
